packages/backend/server/src/core/notification/service.ts

```diff
@@ -0,0 +1,121 @@
+import { Injectable, Logger } from '@nestjs/common';
+
+import {
+  NotificationAccessDenied,
+  NotificationNotFound,
+  PaginationInput,
+} from '../../base';
+import {
+  InvitationNotificationCreate,
+  MentionNotificationCreate,
+  Models,
+  NotificationType,
+  UnionNotification,
+  UnionNotificationBody,
+} from '../../models';
+import { DocReader } from '../doc';
+import { PermissionService } from '../permission';
+
+@Injectable()
+export class NotificationService {
+  private readonly logger = new Logger(NotificationService.name);
+
+  constructor(
+    private readonly models: Models,
+    private readonly permission: PermissionService,
+    private readonly docReader: DocReader
+  ) {}
+
+  async cleanReadExpiredNotifications() {
+    return await this.models.notification.cleanReadExpiredNotifications();
+  }
+
+  async createMention(input: MentionNotificationCreate) {
+    return await this.models.notification.createMention(input);
+  }
+
+  async createInvitation(input: InvitationNotificationCreate) {
+    // is user already a member, skip it
+    const isMember = await this.permission.isWorkspaceMember(
+      input.body.workspaceId,
+      input.userId
+    );
+    if (isMember) {
+      this.logger.debug(
+        `User ${input.userId} is already a member of workspace ${input.body.workspaceId}, skip creating notification`
+      );
+      return;
+    }
+    return await this.models.notification.createInvitation(
+      input,
+      NotificationType.Invitation
+    );
+  }
+
+  async createInvitationAccepted(input: InvitationNotificationCreate) {
+    return await this.models.notification.createInvitation(
+      input,
+      NotificationType.InvitationAccepted
+    );
+  }
+
+  async createInvitationRejected(input: InvitationNotificationCreate) {
+    return await this.models.notification.createInvitation(
+      input,
+      NotificationType.InvitationBlocked
+    );
+  }
+
+  async markAsRead(userId: string, notificationId: string) {
+    const notification = await this.models.notification.get(notificationId);
```

Comment:
the `read-update` round is slow for mark notifications read. do it by `update({ userId, id })` directly. error when marking read is meaning less especially telling them they are marking wrong notifications read